dhall/src/Dhall.hs

```diff
@@ -450,8 +467,10 @@ inputWithSettings settings (Decoder {..}) txt = do
                 bytes' = bytes <> " : " <> suffix
             _ ->
                 Annot expr' expected
-    _ <- Dhall.Core.throws (Dhall.TypeCheck.typeWith (view startingContext settings) annot)
-    let normExpr = Dhall.Core.normalizeWith (view normalizer settings) expr'
+    let substituted = Dhall.Substitution.substitute annot $ view substitutions settings
+    let substituted' = Dhall.Substitution.substitute expr' $ view substitutions settings
+    _ <- Dhall.Core.throws (Dhall.TypeCheck.typeWith (view startingContext settings) substituted)
+    let normExpr = Dhall.Core.normalizeWith (view normalizer settings) substituted'
```

Comment:
I guess that won't work since normalization uses Dhall.Normaliztion.shift which will prevent us from substituting captured variables.
But we can substitute before annotating...